### DISCUSSION TITLE
[@types/node] missing ReadableStream.from static method

### DISCUSSION THREAD

**Discussion by cabralpinto** [DefinitelyTyped/DefinitelyTyped#71154] [Issues with a @types package]
I'm getting the following error when attempting to use `ReadableStream.from`:

> `Property 'from' does not exist on type '{ new (underlyingSource: UnderlyingByteSource, strategy?: { highWaterMark?: number | undefined; } | undefined): ReadableStream<Uint8Array>; new <R = any>(underlyingSource: UnderlyingDefaultSource<...>, strategy?: QueuingStrategy<...> | undefined): ReadableStream<...>; new <R = any>(underlyingSource?: UnderlyingSource...'.ts(2339)`

However, I see that this feature has already been merged (#68861). I tried to update `@types/node` to the latest version (`"^22.9.0"`) to no avail. What could be the issue? Thanks!

[a comment by an automated account is not shown]

**Comment by Renegade334** [comment]
Sounds like you have the DOM library definitions loaded. `ReadableStream.from()` is not yet a stable web standard, and doesn't appear in lib.dom.

If you want to use Node's own types, then either provide a tsconfig `lib` option that doesn't include lib.dom, or explicitly import `ReadableStream` from `node:stream/web`.

**Reply by cabralpinto** [reply]
I tried both options. Removing the `"DOM"` library from `tsconfig.json` does not seem to change anything:

```
"lib": [
  "esnext",
  // "DOM",
  // "DOM.Iterable"
]
```

Explicitly importing `ReadableStream` from `node:stream/web` does fix the error, but then I get another error when trying to pass the stream into a `Response` constructor:

> `The argument of type 'ReadableStream<string>' is not assignable to the parameter of type 'BodyInit | null | undefined'. The type 'ReadableStream<string>' is not assignable to the type 'ReadableStream<any>'. The types of the 'pipeThrough' property are incompatible. The type '<T>(transform: ReadableWritablePair<T, string>, options?: StreamPipeOptions | undefined) => ReadableStream<T>' is not assignable to the type '<T>(transform: ReadableWritablePair<T, any>, options?: StreamPipeOptions | undefined) => ReadableStream<T>'. The types of the parameters 'transform' and 'transform' are incompatible. The type 'ReadableWritablePair<T, any>' is not assignable to the type 'ReadableWritablePair<T | undefined, string>'. The types of the 'readable' property are incompatible. The type 'ReadableStream<T>' is missing the following properties of type 'ReadableStream<T | undefined>': values, [Symbol.asyncIterator]ts(2345)`

What do you suggest?

**Reply by Renegade334** [reply]
So, the compiler is definitely pulling in the web library somewhere. The module import workaround will only work if you're passing `ReadableStream`s to Node APIs, but not if you're passing it to other global web APIs, since the implementations in the web types and @types/node aren't currently compatible with one another.

If you're still finding that tsc is reporting the original error with a tsconfig that explicitly doesn't include lib.dom or lib.webworker (`tsc --showConfig` might be worth checking to make sure), then it might be that a dependency is doing it for you with a reference directive. Difficult to speculate without seeing your project config.

In terms of the issue as a whole, that method isn't missing from the Node types, it's just that your project is loading the web types for whatever reason, and @types/node deliberately doesn't alter those when they're present for compatibility reasons.
